// vike/dist/cjs/node/plugin/plugins/envVars.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.applyEnvVar = exports.envVarsPlugin = void 0;
const vite_1 = require("vite");
const utils_js_1 = require("../utils.js");
const rollupSourceMap_js_1 = require("../shared/rollupSourceMap.js");
const getFilePath_js_1 = require("../shared/getFilePath.js");
function envVarsPlugin() {
    let envsAll;
    let config;
    return {
        name: 'vike:envVars',
        enforce: 'post',
        configResolved(config_) {
            config = config_;
            config.command;
            envsAll = (0, vite_1.loadEnv)(config.mode, config.envDir || config.root, '');
            config.plugins.sort((0, utils_js_1.lowerFirst)((plugin) => (plugin.name === 'vite:define' ? 1 : 0)));
        },
        transform(code, id, options) {
            (0, utils_js_1.assertPosixPath)(id);
            if (id.includes('/node_modules/'))
                return;
            if (!id.startsWith(config.root))
                return;
            if (!code.includes('import.meta.env.'))
                return;
            const isBuild = config.command === 'build';
            const isClientSide = getIsClientSide(config, options);
            Object.entries(envsAll)
                .filter(([key]) => {
                // Already handled by Vite
                const envPrefix = !config.envPrefix ? [] : (0, utils_js_1.isArray)(config.envPrefix) ? config.envPrefix : [config.envPrefix];
                return !envPrefix.some((prefix) => key.startsWith(prefix));
            })
                .forEach(([envName, envVal]) => {
                // Security check
                {
                    const envStatement = getEnvStatement(envName);
                    const publicPrefix = 'PUBLIC_ENV__';
                    const isPrivate = !envName.startsWith(publicPrefix);
                    if (isPrivate && isClientSide) {
                        if (!code.includes(envStatement))
                            return;
                        const modulePath = (0, getFilePath_js_1.getModuleFilePath)(id, config);
                        const errMsgAddendum = isBuild ? '' : ' (Vike will prevent your app from building for production)';
                        const keyPublic = `${publicPrefix}${envName}`;
                        const errMsg = `${envStatement} is used in client-side file ${modulePath} which means that the environment variable ${envName} will be included in client-side bundles and, therefore, ${envName} will be publicly exposed which can be a security leak${errMsgAddendum}. Use ${envStatement} only in server-side files, or rename ${envName} to ${keyPublic}, see https://vike.dev/env`;
                        if (isBuild) {
                            (0, utils_js_1.assertUsage)(false, errMsg);
                        }
                        else {
                            // - Only a warning for faster development DX (e.g. when user toggles `ssr: boolean` or `onBeforeRenderIsomorph: boolean`).
                            // - But only showing a warning can be confusing: https://github.com/vikejs/vike/issues/1641
                            (0, utils_js_1.assertWarning)(false, errMsg, { onlyOnce: true });
                        }
                    }
                    // Double check
                    (0, utils_js_1.assert)(!(isPrivate && isClientSide) || !isBuild);
                }
                // Apply
                code = applyEnvVar(envName, envVal, code);
            });
            // Line numbers didn't change.
            //  - We only break the column number of a couple of lines, wich is acceptable.
            //  - Anyways, I'm not even sure Vite supports high-resolution column number source mapping.
            const ret = (0, rollupSourceMap_js_1.sourceMapPassthrough)(code);
            return ret;
        }
    };
}
exports.envVarsPlugin = envVarsPlugin;
function applyEnvVar(envName, envVal, code) {
    const envStatement = getEnvStatement(envName);
    const regex = new RegExp((0, utils_js_1.escapeRegex)(envStatement) + '\\b', 'g');
    return code.replace(regex, JSON.stringify(envVal));
}
exports.applyEnvVar = applyEnvVar;
function getEnvStatement(envName) {
    return `import.meta.env.${envName}`;
}
function getIsClientSide(config, options) {
    const isBuild = config.command === 'build';
    if (isBuild) {
        (0, utils_js_1.assert)(typeof config.build.ssr === 'boolean');
        const isServerSide = config.build.ssr;
        if (options !== undefined) {
            (0, utils_js_1.assert)(options.ssr === isServerSide);
        }
        return !isServerSide;
    }
    else {
        (0, utils_js_1.assert)(config.build.ssr === false);
        (0, utils_js_1.assert)(typeof options?.ssr === 'boolean');
        const isServerSide = options.ssr;
        return !isServerSide;
    }
}
