// vike/dist/cjs/node/runtime/globalContext.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getRuntimeManifest = exports.setGlobalContext_prerender = exports.setGlobalContext_viteDevServer = exports.getViteConfig = exports.getViteDevServer = exports.getGlobalContext = exports.initGlobalContext = exports.getGlobalContextAsync = exports.getGlobalContextSync = void 0;
const utils_js_1 = require("./utils.js");
const loadImportBuild_js_1 = require("./globalContext/loadImportBuild.js");
const getPageFiles_js_1 = require("../../shared/getPageFiles.js");
const assertPluginManifest_js_1 = require("../shared/assertPluginManifest.js");
const getConfigVike_js_1 = require("../shared/getConfigVike.js");
const assertRuntimeManifest_js_1 = require("../shared/assertRuntimeManifest.js");
const picocolors_1 = __importDefault(require("@brillout/picocolors"));
let resolveGlobalContext;
const globalObject = (0, utils_js_1.getGlobalObject)('globalContext.ts', {
    globalContextPromise: new Promise((r) => (resolveGlobalContext = r))
});
function getGlobalContext() {
    (0, utils_js_1.assert)(globalObject.globalContext);
    return globalObject.globalContext;
}
exports.getGlobalContext = getGlobalContext;
/** @experimental https://vike.dev/getGlobalContext */
function getGlobalContextSync() {
    (0, utils_js_1.assertUsage)(globalObject.globalContext, "The global context isn't set yet, call getGlobalContextSync() later or use getGlobalContextAsync() instead.");
    return makePublic(globalObject.globalContext);
}
exports.getGlobalContextSync = getGlobalContextSync;
/** @experimental https://vike.dev/getGlobalContext */
async function getGlobalContextAsync() {
    await globalObject.globalContextPromise;
    (0, utils_js_1.assert)(globalObject.globalContext);
    return makePublic(globalObject.globalContext);
}
exports.getGlobalContextAsync = getGlobalContextAsync;
function makePublic(globalContext) {
    const globalContextPublic = {
        assetsManifest: globalContext.assetsManifest
    };
    // Add internals (and prepended _ prefix to their keys)
    {
        const publicKeys = Object.keys(globalContextPublic);
        (0, utils_js_1.objectKeys)(globalContext)
            .filter((key) => !publicKeys.includes(key))
            .forEach((key) => {
            const keyPublic = `_${key}`;
            Object.defineProperty(globalContextPublic, keyPublic, {
                enumerable: true,
                get() {
                    (0, utils_js_1.assertWarning)(false, `Using internal globalContext.${keyPublic} which is discouraged: it may break in any minor version update. Instead, reach out on GitHub and elaborate your use case.`, {
                        onlyOnce: true
                    });
                    return globalContext[key];
                }
            });
        });
    }
    return globalContextPublic;
}
function setGlobalContext_viteDevServer(viteDevServer) {
    if (globalObject.viteDevServer)
        return;
    (0, utils_js_1.assert)(!globalObject.globalContext);
    (0, utils_js_1.assert)(!globalObject.globalContext);
    globalObject.viteConfig = viteDevServer.config;
    globalObject.viteDevServer = viteDevServer;
}
exports.setGlobalContext_viteDevServer = setGlobalContext_viteDevServer;
function getViteDevServer() {
    return globalObject.viteDevServer ?? null;
}
exports.getViteDevServer = getViteDevServer;
function setGlobalContext_prerender(viteConfig) {
    if (globalObject.viteConfig)
        return;
    (0, utils_js_1.assert)(!globalObject.globalContext);
    globalObject.viteConfig = viteConfig;
}
exports.setGlobalContext_prerender = setGlobalContext_prerender;
function getViteConfig() {
    return globalObject.viteConfig ?? null;
}
exports.getViteConfig = getViteConfig;
async function initGlobalContext(isPrerendering = false, outDir) {
    if (globalObject.globalContext)
        return;
    const { viteDevServer, viteConfig } = globalObject;
    (0, utils_js_1.assertNodeEnv_runtime)(!!viteDevServer);
    const isProduction = !viteDevServer;
    if (!isProduction) {
        (0, utils_js_1.assert)(viteConfig);
        (0, utils_js_1.assert)(!isPrerendering);
        const configVike = await (0, getConfigVike_js_1.getConfigVike)(viteConfig);
        const pluginManifest = getRuntimeManifest(configVike);
        globalObject.globalContext = {
            isProduction: false,
            isPrerendering: false,
            assetsManifest: null,
            pluginManifest: null,
            viteDevServer,
            viteConfig,
            baseServer: pluginManifest.baseServer,
            baseAssets: pluginManifest.baseAssets,
            includeAssetsImportedByServer: pluginManifest.includeAssetsImportedByServer,
            redirects: pluginManifest.redirects,
            trailingSlash: pluginManifest.trailingSlash,
            disableUrlNormalization: pluginManifest.disableUrlNormalization
        };
    }
    else {
        const buildEntries = await (0, loadImportBuild_js_1.loadImportBuild)(outDir);
        assertBuildEntries(buildEntries, isPrerendering ?? false);
        const { pageFiles, assetsManifest, pluginManifest } = buildEntries;
        (0, getPageFiles_js_1.setPageFiles)(pageFiles);
        assertViteManifest(assetsManifest);
        (0, assertPluginManifest_js_1.assertPluginManifest)(pluginManifest);
        const globalContext = {
            isProduction: true,
            assetsManifest,
            pluginManifest,
            viteDevServer: null,
            baseServer: pluginManifest.baseServer,
            baseAssets: pluginManifest.baseAssets,
            includeAssetsImportedByServer: pluginManifest.includeAssetsImportedByServer,
            redirects: pluginManifest.redirects,
            trailingSlash: pluginManifest.trailingSlash,
            disableUrlNormalization: pluginManifest.disableUrlNormalization
        };
        if (isPrerendering) {
            (0, utils_js_1.assert)(viteConfig);
            const configVike = await (0, getConfigVike_js_1.getConfigVike)(viteConfig);
            (0, utils_js_1.assert)(configVike);
            (0, utils_js_1.objectAssign)(globalContext, {
                isPrerendering: true,
                viteConfig
            });
            globalObject.globalContext = globalContext;
        }
        else {
            (0, utils_js_1.objectAssign)(globalContext, {
                isPrerendering: false,
                viteConfig: null
            });
            globalObject.globalContext = globalContext;
        }
    }
    resolveGlobalContext(globalObject.globalContext);
}
exports.initGlobalContext = initGlobalContext;
function getRuntimeManifest(configVike) {
    const { includeAssetsImportedByServer, baseServer, baseAssets, redirects, trailingSlash, disableUrlNormalization } = configVike;
    const manifest = {
        baseServer,
        baseAssets,
        includeAssetsImportedByServer,
        redirects,
        trailingSlash,
        disableUrlNormalization
    };
    (0, assertRuntimeManifest_js_1.assertRuntimeManifest)(manifest);
    return manifest;
}
exports.getRuntimeManifest = getRuntimeManifest;
function assertBuildEntries(buildEntries, isPreRendering) {
    const errMsg = [
        `You are tyring to run`,
        isPreRendering ? 'pre-rendering' : 'the server for production',
        `but your app isn't built yet. Run ${picocolors_1.default.cyan('$ vite build')} before `,
        isPreRendering ? 'pre-rendering.' : 'running the server.'
    ].join(' ');
    (0, utils_js_1.assertUsage)(buildEntries, errMsg);
}
function assertViteManifest(manifest) {
    (0, utils_js_1.assert)((0, utils_js_1.isPlainObject)(manifest));
    /* We should include these assertions but we don't as a workaround for PWA manifests: https://github.com/vikejs/vike/issues/769
       Instead, we should rename the vite manifest e.g. with https://vitejs.dev/config/build-options.html#build-manifest
    Object.entries(manifest)
      // circumvent esbuild bug: esbuild adds a `default` key to JSON upon `require('./some.json')`.
      .filter(([key]) => key !== 'default')
      .forEach(([_, entry]) => {
        assert(isPlainObject(entry))
        assert(typeof entry.file === 'string')
      })
    */
}
